Slack: New Command: Send Message (ISSUE-15424)

RobErskine opened the pull request:
## Description

This is for feature request #15424, adding in the ability to send instant and scheduled messages.  

- Adds a new command called "Send Message". In this command you can select your recipient (either user or channel), add your message, select when the message should be sent (defaults to now, but can be scheduled for later).
- Updates the README with instructions on how to add the necessary scopes to get this new command to work.

## Screencast
https://github.com/user-attachments/assets/896a07d2-8b31-4f62-a645-6ce442d843ae

_Sorry for the sniffles, I'm getting over a cold :(_ 

## Checklist

- [x] I read the [extension guidelines](https://developers.raycast.com/basics/prepare-an-extension-for-store)
- [x] I read the [documentation about publishing](https://developers.raycast.com/basics/publish-an-extension)
- [x] I ran `npm run build` and [tested this distribution build in Raycast](https://developers.raycast.com/basics/prepare-an-extension-for-store#metadata-and-configuration)
![sionsextensionsslack-2025-01-23-22 09 17@2x](https://github.com/user-attachments/assets/551ea762-b37f-4c66-8f4a-58516de55325)

- [x] I checked that files in the `assets` folder are used by the extension itself
- [x] I checked that assets used by the `README` are placed outside of the `metadata` folder

pernielsentikaer commented:
Thanks for your contribution here 🔥 

@momme-rtf did we deal with new permissions before, and can you check this 🙂

RobErskine commented:
I'm re-using the `handleError(error)` function used elsewhere in the extension. Before I knew that existed though, I had something like this...

```
if (error.message.contains('missing_scope') { 
  return ( 
    <Detail markdown={`
      # Missing Scopes \n 
      To use this command, please ensure that you have \`chat:write\` scope enabled. 
      Please add the required scopes and re-install the Slack app.
    `} />
  )
}
```

This would provide instructions on what users would need for this particular command, instead of just a generic "missing scopes error". If y'all would like I could do something similar with the `handleError(error, 'command-name')` where `command-name` will change the markdown in the Detail to show the user exactly which scopes they need for the command they are trying to use? just a thought!

pernielsentikaer commented:
Would you be able to test it @momme-rtf 

@RobErskine makes sense :)

RobErskine commented:
heya @momme-rtf & @thomaslombart, is there anything I can do to help verify or test this? i've been using the local command in my daily flow and it's been pretty useful. Would love to get it in other people's hands!

pernielsentikaer commented:
@momme-rtf are you still around

pernielsentikaer commented:
@momme-rtf last ping 🙂

RobErskine commented:
For what it's worth, I think the holdup on this PR was adjusting the scope permissions within Slack to get this to work. The scope was just updated in this other PR: https://github.com/raycast/extensions/pull/17586. If there is anything we can learn from that PR to account for her, let me know!

pernielsentikaer commented:
Hi @RobErskine 

Seems like we'll continue without @momme-rtf at the moment, could you correct the merge conflicts?

RobErskine commented:
Thanks @pernielsentikaer. I've gone ahead and fixed the merge conflicts. I also updated the the success state of sending/scheduling a message to use the `showHUD` instead so it pops a user out of the send message view. If you have any other feedback or need help testing or anything please let me know.

pernielsentikaer commented:
@thomaslombart have you added new rights before?

thomaslombart commented:
@pernielsentikaer what do you mean?

RobErskine commented:
@pernielsentikaer @thomaslombart I think we are talking about User Token Scopes in the Slack App settings to give the app more permissions on the user's behalf. In order to use this new "Send Message" command, an admin of the Slack App would need to add the `users:write` scope. This was just done [in another PR that was merged](https://github.com/raycast/extensions/pull/17586/files) in not too long ago. Similar to this PR i've added instructions into the README and Changelog as well. If there is anything else needed please let me know!

![Arc_Slack API Applications  Stellar Central Slack_2025-04-08 3@2x](https://github.com/user-attachments/assets/976a2bc3-4872-4c3d-b51e-1efff7f8be34)

pernielsentikaer commented:
Would it make sense to have a checkbox to `Send now` and if you untick you can schedule?

![slack 2025-04-09 at 09 55 22](https://github.com/user-attachments/assets/8d66065a-bda8-46d3-8d60-278ed8c389d2)

RobErskine commented:
Great suggestion @pernielsentikaer; that cleans up the UI and takes the cognitive load out of when the message will be sent. I just added that in. 

Also added in user and channel icons to the Dropdown as well. Quick video attached. Please feel free to test for yourself and let me know if you have any questions or concerns.

https://github.com/user-attachments/assets/2467dc21-d923-4c15-9837-5ddc75802179

pernielsentikaer commented on extensions/slack/CHANGELOG.md:
```diff
@@ -1,6 +1,9 @@
 # Slack Changelog
 
-## [Moved contributor] - 2025-03-31
+## [New "Send Message" command] - {PR_MERGE_DATE}
```
Did you delete `## [Moved contributor] - 2025-03-31`

RobErskine replied:
If I did it was an accident... my mistake. Just added it back in.

pernielsentikaer reviewed (approved):
Hi 👋 

Looks good to me, approved 🔥
